trace/trace.go

```diff
@@ -0,0 +1,348 @@
+// field added to it. These fields are also passed along to downstream services.
+// It is useful to add fields here that pertain to the entire trace, to aid in
+// filtering spans at many different areas of the trace together.
+func (t *Trace) AddField(key string, val interface{}) {
+	t.tlfLock.Lock()
+	defer t.tlfLock.Unlock()
+	if t.traceLevelFields != nil {
+		t.traceLevelFields[key] = val
+	}
+}
+
+// GetRootSpan returns the root of the in-process trace. Finishing the root span
+// will send the entire trace to Honeycomb. From the root span you can walk the
+// entire span tree using GetChildren.
+func (t *Trace) GetRootSpan() *Span {
+	return t.rootSpan
+}
+
+// Send will finish and send the all synchronous spans in the trace to Honeycomb
+func (t *Trace) Send() {
+	// TODO add sampling
+	// make sure all sync spans are finished
+	rs := t.rootSpan
+	if !rs.amFinished {
+		rs.Finish()
+	}
+	// start at the root span and send them all
+	recursiveSend(rs)
+}
+
+// Span represents a specific task or portion of an application. It has a time
+// and duration, and is linked to parent and children.
+type Span struct {
+	amAsync      bool
+	amFinished   bool
+	amRoot       bool
+	children     []*Span
+	ev           *libhoney.Event
+	spanID       string
+	parentID     string
+	parent       *Span
+	rollupFields map[string]float64
+	rollupLock   sync.Mutex
+	sent         bool // records whether this span has already been sent.
+	timer        timer.Timer
+	trace        *Trace
+}
+
+// newSpan takes care of *some* of the initialization necessary to create a new
+// span. IMPORTANT it is not all of the initialization! It does *not* set parent
+// ID or assign the pointer to the trace that contains this span. See existing
+// uses of this function to get an example of the other things necessary to
+// create a well formed span.
+func newSpan() *Span {
+	return &Span{
+		spanID:       uuid.Must(uuid.NewRandom()).String(),
+		timer:        timer.Start(),
+		children:     make([]*Span, 0),
+		rollupFields: make(map[string]float64),
+	}
+}
+
+// AddField adds a key/value pair to this span
+func (s *Span) AddField(key string, val interface{}) {
+	if s.ev != nil {
+		s.ev.AddField(key, val)
+	}
+}
+
+// AddRollupField adds a key/value pair to this span. If it is called repeatedly
+// on the same span, the values will be summed together.  Additionally, this
+// field will be summed across all spans and added to the trace as a total. It
+// is especially useful for doing things like adding the duration spent talking
+// to a specific external service - eg database time. The root span will then
+// get a field that represents the total time spent talking to the database from
+// all of the spans that are part of the trace.
+func (s *Span) AddRollupField(key string, val float64) {
+	s.rollupLock.Lock()
+	if s.rollupFields != nil {
+		s.rollupFields[key] += val
+	}
+	s.rollupLock.Unlock()
+	if s.trace != nil {
+		s.trace.rollupLock.Lock()
```

Comment:
feels wrong to have spans locking the mutex in the trace.  Why not `Trace.addRollup(key, val)` and have the locks be local to the trace (as in `Trace.AddField`).

Reply:
👍 added rollup method to the trace (and used it here) so its locks stay private.